packages/core/src/shared/performance/performance.ts

```diff
@@ -0,0 +1,118 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import assert from 'assert'
+import { getLogger } from '../logger'
+import { isWeb } from '../extensionGlobals'
+
+interface PerformanceMetrics {
+    cpuUsage: number
+    heapTotal: number
+    duration: number
+}
+
+interface TestOptions {
+    darwin?: Partial<PerformanceMetrics>
+    win32?: Partial<PerformanceMetrics>
+    linux?: Partial<PerformanceMetrics>
+}
+
+export interface PerformanceSpan<T> {
+    value: T
+    performance: PerformanceMetrics
+}
+
+export class PerformanceTracker {
+    #startPerformance:
+        | {
+              cpuUsage: NodeJS.CpuUsage
+              memory: number
+              duration: [number, number]
+          }
+        | undefined
+
+    constructor(private readonly name: string) {}
+
+    static enabled(name: string, trackPerformance?: boolean): boolean {
+        return name === 'function_call' && (trackPerformance ?? false) && !isWeb()
+    }
+
+    start() {
+        this.#startPerformance = {
+            cpuUsage: process.cpuUsage(),
+            memory: process.memoryUsage().heapTotal,
+            duration: process.hrtime(),
+        }
+    }
+
+    stop(): PerformanceMetrics | undefined {
+        if (this.#startPerformance) {
+            const endCpuUsage = process.cpuUsage(this.#startPerformance?.cpuUsage)
+            const userCpuUsage = endCpuUsage.user / 1000000
+            const systemCpuUsage = endCpuUsage.system / 1000000
+
+            const elapsedTime = process.hrtime(this.#startPerformance.duration)
+            const duration = elapsedTime[0] + elapsedTime[1] / 1e9 // convert microseconds to seconds
+            const usage = ((userCpuUsage + systemCpuUsage) / duration) * 100 // convert to percentage
+
+            const endMemoryUsage = process.memoryUsage().heapTotal - this.#startPerformance?.memory
+            const endMemoryUsageInMB = endMemoryUsage / (1024 * 1024) // converting bytes to MB
+
+            return {
+                cpuUsage: usage,
+                heapTotal: endMemoryUsageInMB,
+                duration,
+            }
+        } else {
+            getLogger().debug(`PerformanceTracker: start performance not defined for ${this.name}`)
+        }
+    }
+}
+
+export function performanceTest(options: TestOptions, name: string, fn: () => Promise<void>): Mocha.Test
+export function performanceTest(options: TestOptions, name: string, fn: () => void): Mocha.Test
+export function performanceTest(options: TestOptions, name: string, fn: () => void | Promise<void>) {
+    const testOption = options[process.platform as 'linux' | 'darwin' | 'win32']
+
+    const performanceTracker = new PerformanceTracker(name)
+
+    return it(name, async () => {
```

Comment:
Just to double-check... we didn't (and won't) promote mocha to a "production" dependency, right?

Reply:
no, this is only for tests. Ideally telemetry tells us the rough runtime of functions then we can use that data to reduce the options in the performance test. That way we can see regressions as well